### PR TITLE
Disable label name validation

### PR DESCRIPTION
**Pull request by trevorwhitney**
**What this PR does**:
This PR enables users who want to store non-conventional prometheus data in cortex by exposing the ability to disable label name validation via a request parameter. This option is currently available to in the distributor [config](https://github.com/cortexproject/cortex/blob/eb02c0dc3bd9c4010baa9c77b84647111b1a013b/pkg/distributor/distributor.go#L174), but that requires you to run a whole separate distributor to handle requests for this type. By exposing this option through the `WriteRequest`, a single distributor can handle both payload types.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/distributor/distributor_test.go
Comment (jtlisi) [inline pkg/distributor/distributor_test.go:7]: This import needs to be separate from the standard library imports. We run rather strict linter settings for the project and this is triggering the CI failure.

**Review by jtlisi — APPROVED** [review]
LGTM, modulo linting fix

**Comment by roidelapluie** [comment]
What is the usecase here? I expect promql not to work without proper metric names.

**Comment by roidelapluie** [comment]
I also quote @pstibrany 

> > Now the question is, do we want to be a generic TSDB? I don't know, I want it to be.
>
> I don't think Cortex should try to be a generic TSDB. There are already projects that do that better, and Cortex' strength is in making Prometheus horizontally-scalable (wherever that eventually leads). If there is low-hanging fruit to make some use-cases easier, I'm all for it, but generic TSDB is completely different beast altogether (starting with query language, underlying storage, architecture choices).

https://github.com/cortexproject/cortex/pull/3708

**Comment by pstibrany** [comment]
> What is the usecase here? I expect promql not to work without proper metric names.

As far as I understand it, the use case here is to add a for a non-Prometheus data into Cortex. This would be achieved by 1) having a proxy for incoming data that would sit in front of distributor, and call Push on distributor via gRPC with this new flag set to true (skip validation). 2) On query time, there would be a special querier that doesn't use PromQL at all, but use Cortex' store-gateway to fetch required data.

Additional support in Cortex needed to make this work is pretty small: disable label validation (this PR), and expose distributor push via gRPC. I think this falls well within "low-hanging fruit to make some use-cases easier".

File: pkg/ingester/client/cortex.proto
Comment (pstibrany) [inline pkg/ingester/client/cortex.proto:36]: I'm wondering if we should use `google.protobuf.Any hints` approach here. Strictly speaking this isn't supported by Ingester service, so it doesn't belong here. But Distributor implements the same method. (When we expose Distributor via gRPC, I expect it will reuse the same `Push` method, with `cortex.WriteRequest` and `cortex.WriteResponse`)

File: pkg/distributor/distributor.go
Comment (pstibrany) [inline pkg/distributor/distributor.go:369]: Nice catch! I've sent separate PR #3738 to fix double-import of `ingester_client` package in this file.

File: pkg/ingester/client/cortex.proto
Comment (pracucci) [reply pkg/ingester/client/cortex.proto:36]: We need to keep `WriteRequest` compatible with Prometheus one. If we use the field number `4`, it's very likely it will clash next time Prometheus remote write adds a new field. To fix this, we could use a very high field number (eg. `1000`).

File: pkg/distributor/distributor.go
Comment (pracucci) [reply pkg/distributor/distributor.go:369]: We've just merged https://github.com/cortexproject/cortex/pull/3738. Could you rebase `master`, please?

File: pkg/distributor/distributor.go
Comment (pracucci) [inline pkg/distributor/distributor.go:498]: @replay `SkipLabelNameValidation` was added to the config in https://github.com/cortexproject/cortex/pull/2835, based on needs of a downstream project. Do we still need it in the config or can we remote it favour of specifying it directly in the request? If it's needed in the config too, then OK, otherwise I would prefer to keep code cleaner.

Comment (replay) [reply pkg/distributor/distributor.go:498]: You're right, we won't need the `SkipLabelNameValidation` in the config struct anymore. 

Would it be fine for you if we remove that flag from the Distributor config in a separate PR after this one has been merged? That would help because otherwise merging this PR is going to break the down-stream CI tests of projects relying on the Distributor config flag, if we remove it in a separate PR we can first switch the down-stream users to rely on this new per-request flag and only after that we remove the Distributor config flag.

Comment (pracucci) [reply pkg/distributor/distributor.go:498]: I'm fine if you will do it in a separate PR.

File: pkg/distributor/distributor_test.go
Comment (trevorwhitney) [reply pkg/distributor/distributor_test.go:7]: ahh, good to know, thanks!

File: pkg/ingester/client/cortex.proto
Comment (trevorwhitney) [reply pkg/ingester/client/cortex.proto:36]: @pstibrany is the `google.protobuf.Any hints` approach like the `options` field in the rules proto (https://github.com/cortexproject/cortex/blob/d62514185ca7678f0e77ab1a1f93d291409fd530/pkg/ruler/rules/rules.proto#L29), making this field extensible in the case of future config options? Or am I missing the boat here?

**Comment by trevorwhitney** [comment]
Thanks everyone, just pushed up a commit that hopefully resolves above mentioned issues.

* created a new grpc service with just the `Push` method to expose the distributor via grpc
* changed the grpc field position in `WriteRequest` to `1000` to guard against future changes to upstream `WriteRequest` in prometheus
* updated the `util/push/push.go` to always set the `SkipLabelNameValidation` property to `false` so that all http request will continue to be validated.

**Review by replay — APPROVED** [review]
LGTM

(trevor, not that I'm not a maintainer, so my approval doesn't count towards the requirement of `2` to merge)

**Comment by trevorwhitney** [comment]
haha, @replay appreciate the support nonetheless!

**Review by pstibrany — APPROVED** [review]
LGTM.

(Nit: it seems little strange that "cortex.Distributor" gRPC service is defined in ingester package)

**Comment by trevorwhitney** [comment]
@pstibrany how would you feel about renaming the new service to `PushIngestor` instead of `Distrubutor`, since that's really what it is? That way it would probably make more sense to keep in the ingester package.

**Comment by pstibrany** [comment]
> @pstibrany how would you feel about renaming the new service to `PushIngestor` instead of `Distrubutor`, since that's really what it is? That way it would probably make more sense to keep in the ingester package.

I don’t have a strong preference, and also I’m fine with keeping it as is.

**Comment by trevorwhitney** [comment]
I went ahead with the rename of `Distributor` service to `PushOnlyIngester` as I think (as @pstibrany pointed out) that makes more sense for something in the ingester package. I think this is ready for a merge, but I don't have write permission to do so. Thanks!

**Review by pracucci — APPROVED** [review]
LGTM! Please remind to open a PR to remove the "skip label name validation" from config, since you mentioned you don't need to anymore.

File: pkg/distributor/distributor_test.go
Comment (replay) [reply pkg/distributor/distributor_test.go]: just a bit more detail about the linting: https://cortexmetrics.io/docs/contributing/#formatting

**Pull request merged**
